compiler/ast/ast.go

```diff
@@ -756,3 +763,35 @@ type Agg struct {
 	Where    Expr   `json:"where"`
 	Loc      `json:"loc"`
 }
+
+// ----------------------------------------------------------------------------
+// Functions
+//
+// FnDecl (defined above) binds a lambda to a name that can be called
+// FnLamba is a nameless body with formal parameters that can be called or referenced
+// FnName is a reference to function's name for calling or passing to an op or map-call
+// FnRef is a sum type representing FnName or FnLambda (a named fn reference or a lambda value)
+
+type FnLambda struct {
+	Kind    string `json:"kind" unpack:""`
+	Formals []*ID  `json:"formals"`
+	Expr    Expr   `json:"expr"`
+	Loc     `json:"loc"`
+}
+
+type FnName struct {
+	Kind string `json:"kind" unpack:""`
+	ID   *ID    `json:"id"`
```

Comment:
Nit: Is there any upside to using ID here instead of string?

```suggestion
	Text string `json:"text"`
```